akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -199,25 +221,45 @@ object ShardCoordinator {
    * 10 shards per node. One node may have 19 shards and others 10 without a rebalance occurring.
    *
    * The number of ongoing rebalancing processes can be limited by `maxSimultaneousRebalance`.
+   *
+   * During a rolling upgrade (when nodes with multiple application versions are present) allocating to
+   * old nodes are avoided.
+   *
+   * Not intended for user extension.
    */
   @SerialVersionUID(1L)
+  @DoNotInherit
   class LeastShardAllocationStrategy(rebalanceThreshold: Int, maxSimultaneousRebalance: Int)
-      extends ShardAllocationStrategy
+      extends ActorSystemDependentAllocationStrategy
       with Serializable {
 
+    import LeastShardAllocationStrategy._
+
+    @volatile private var cluster: Cluster = _
+
+    override def start(system: ActorSystem): Unit = {
+      cluster = Cluster(system)
+    }
+
+    protected def rollingUpdateInProgress: Boolean =
+      cluster.state.hasMoreThanOneAppVersion
+
     override def allocateShard(
         requester: ActorRef,
         shardId: ShardId,
         currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]]): Future[ActorRef] = {
-      val (regionWithLeastShards, _) = currentShardAllocations.minBy { case (_, v) => v.size }
+      val (regionWithLeastShards, _) = possibleShardDestinations(currentShardAllocations).minBy {
+        case (_, v) => v.size
+      }
       Future.successful(regionWithLeastShards)
     }
 
     override def rebalance(
         currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]],
         rebalanceInProgress: Set[ShardId]): Future[Set[ShardId]] = {
-      if (rebalanceInProgress.size < maxSimultaneousRebalance) {
+      if (rebalanceInProgress.size < maxSimultaneousRebalance && !rollingUpdateInProgress) {
```

Comment:
I wonder if some users would deploy a new node with new version without shutting down old node. Like a canary deployment. Testing one new before rolling out fully. Then this would mean that such new node would not be used. We have a ticket for #29542 where we might consider this.